### PR TITLE
add ndjson fuzzer

### PR DESCRIPTION
**Pull request by pauldreik**
This adds a fuzzer for ndjson. It will find stuff, reported in https://github.com/simdjson/simdjson/pull/1296. This PR replaces that one, as it contains only the ndjson fuzzer plus removal of the deprecated cmake GIT option.

### CODE REVIEW

File: fuzz/fuzz_ndjson.cpp [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#endif
+    simdjson::dom::document_stream docs;
+    if(parser.parse_many(json,batch_size).get(docs)) {
+      return 0;
+    }
+
+  size_t bool_count=0;
+  for (auto doc : docs) {
+    bool_count+=doc.is_bool();
+  }
```
Comment (pauldreik) [inline fuzz/fuzz_ndjson.cpp:25]: Is this sufficient to detect https://github.com/simdjson/simdjson/issues/1302 ?

Comment (lemire) [reply fuzz/fuzz_ndjson.cpp:25]: I would not worry about 1302.

Comment (lemire) [reply fuzz/fuzz_ndjson.cpp:25]: But if you want to check for 1302, then just put an empty loop.

**Comment by lemire** [comment]
@pauldreik Would you sync your branch with the main simdjson branch? I suspect that the issues may go away.

**Comment by pauldreik** [comment]
Sorry that I have not had time looking at this yet, busy with other 
stuff! Will probably have a look during the weekend.

**Comment by pauldreik** [comment]
Current status: at least one issue remains (#1310). The fuzzer finds that over and over again, which more or less stops the fuzzer from looking for something else.

**Comment by pauldreik** [comment]
With the fix in https://github.com/simdjson/simdjson/pull/1321 the ndjson fuzzer does not find stuff anymore. As soon as that is merged, this branch should be ready for merge (modulo unit test conflicts).

**Comment by lemire** [comment]
@pauldreik Can you sync with master?

**Comment by pauldreik** [comment]
Synced, and I test ran it shortly both with and without threads without any issues.

First one that sees it passes CI merges!

**Comment by lemire** [comment]
Important work.

After it passes, I think we should merge and release (0.7).

**Comment by pauldreik** [comment]
Unfortunately there is more to do - I made a quick run with thread sanitizers and it reported errors. Have not had time to look at it, but will push a branch so you can see it for yourself.

**Comment by pauldreik** [comment]
Have a look here: https://github.com/simdjson/simdjson/pull/1322

**Comment by lemire** [comment]
Hopefully, thread sanity is recovered with PR https://github.com/simdjson/simdjson/pull/1323

**Comment by lemire** [comment]
@pauldreik 

Can you resync with master?

**Comment by pauldreik** [comment]
I synced with master locally just now (got 
59c857e969e9822f3bb5fd67aea13a3c6c8cbefe), and it did unfortunately not 
help the thread sanitizer errors. But let's not keep that from merging 
the ndjson fuzzer - it is more important to get the ndjson fuzzer to 
oss-fuzz, so it can fuzz it more thorough than I do locally.

**Comment by lemire** [comment]
Ok. Merging.

**Comment by lemire** [comment]
@pauldreik Is your expectation that now that we have merged this PR (again, important work), we should be getting bug warnings from oss-fuzz? Or did we get some and I just missed them? Or do we need to activate something?

I am asking because I want to release 0.7 soon and I don't want to release with outstanding bugs that have been flagged.

We are pretty much ready to release 0.7. A core new feature is your fuzzing. (Well, it is not really a feature per se, but you see what I mean.) It implies that the fuzzing has run.

**Comment by pauldreik** [comment]
I think we should allow for oss-fuzz at least 48 hours (up to 24 hours 
until it is built, then 24 hours fuzzing). No need to do anything 
besides merging. Inspecting the coverage results (usually shows up in 
the afternoon UTC) should give us a hint how much the new fuzzer ran.

I do not expect anything to show up, but you never know... oss-fuzz does 
for instance not have the 4k input size limit I apply in CI.

There is also the powerpc fuzzing that needs to happen. I could give it 
a go hopefully tomorrow, unless you want to have a stab at it. It should 
be as easy as the already existing arm64 fuzzing which is already in CI.

I am a fan of frequent releases, so I think it is better to go for 0.7 
than waiting for us to understand and fix the thread sanitizer issues!

Den 2020-12-02 kl. 17:36, skrev Daniel Lemire:
> @pauldreik Is your expectation that now that we have merged this PR (again, important work), we should be getting bug warnings from oss-fuzz? Or did we get some and I just missed them? Or do we need to activate something?
> 
> I am asking because I want to release 0.7 soon and I don't want to release with outstanding bugs that have been flagged.
> 
> We are pretty much ready to release 0.7. A core new feature is your fuzzing. (Well, it is not really a feature per se, but you see what I mean.) It implies that the fuzzing has run.
> 
> 
>

**Comment by lemire** [comment]
I think we can wait two days to see if something comes up. If nothing comes up, I will be happy to push version 0.7.

(I hate patch releases and much prefer not to have to backtrack immediately after a minor release.)

**Pull request merged**